### ISSUE TITLE
[F03] bogus error: Could not resolve generic type bound procedure

### ISSUE THREAD

**Issue by janusw** [flang-compiler/flang#533]
Test case:

```
module m

implicit none

type :: s
   character(len=:), allocatable :: cs
contains
   procedure, nopass :: a1
   procedure         :: a2
   generic           :: a => a1, a2
end type

contains

type(s) function a1(cs)
   character(len=*), intent(in) :: cs
end function

type(s) function a2(self)
   class(s), intent(in) :: self
   if (allocated(self%cs)) then
      a2 = self%a(self%cs)
   else
      a2 = self%a('')
   end if
end function

end module
```

Although this is supposed to be perfectly valid Fortran 2003 code, flang complains:
```
F90-S-0155-Could not resolve generic type bound procedure a (test.f90: 22)
F90-S-0155-Could not resolve generic type bound procedure a (test.f90: 22)
F90-S-0148-Reference to derived type required (test.f90: 22)
F90-S-0155-Could not resolve generic type bound procedure a (test.f90: 24)
F90-S-0155-Could not resolve generic type bound procedure a (test.f90: 24)
F90-S-0148-Reference to derived type required (test.f90: 24)
  0 inform,   0 warnings,   6 severes, 0 fatal for a2
```

**Comment by janusw** [comment]
As usual, this bug is shared with pgfortran 18.04.

For me personally, it is currently the number-one blocker that keeps me from using flang.
